tests/functional/test_install.py

```diff
@@ -1391,6 +1391,50 @@ def test_install_pep508_with_url_in_install_requires(script):
     assert "Successfully installed packaging-15.3" in str(res), str(res)
 
 
+@pytest.mark.parametrize('create_dep, format', [
+    (create_test_package_with_setup, 'directory'),
+    (create_basic_wheel_for_package, 'wheel'),
```

Comment:
Adding these strings "directory" and "wheel" and then doing if-checks in the test method to toggle behavior is a sign that this test would be better as two separate tests. To share code, you can put the common code in a private helper method used by both (e.g. still passing `create_dep`).

Reply:
I went back and forth on that originally. I couldn't come up with a helper that makes sense, so I went with parametrization and the conditional.

Pushed up a version where it's split into two tests though

Reply:
> I couldn't come up with a helper that makes sense,

For the helper, you can basically use the body of the test function you had before more or less as is -- minus the `if format` toggles at the beginning and the end. Then each new test will have the form `<beginning> <helper call> <end>`, where the beginning and end portions are the if clauses that would have executed in your original test. There's a lot of repetition in what you have now in a long-ish test, so it would be helpful in understanding to be able to see the common structure being used.

Reply:
This seems a lot harder to follow than the original parametrized version:

```python
def function_that_does_a_lot(script, create_dep, dep_v1_path):
    dep_v2_path = create_dep(script, name='dep', version='2.0')

    pkga_path = create_basic_wheel_for_package(
        script, name='pkga', version='1.0',
        depends=['dep @ ' + path_to_url(dep_v1_path)],
    )
    res = script.pip('install', pkga_path)
    assert "Successfully installed dep-1.0" in str(res), str(res)

    pkga_path.rmtree()

    # Updating the URL to the dependency installs the updated dependency
    pkga_path = create_basic_wheel_for_package(
        script, name='pkga', version='2.0',
        depends=['dep @ ' + path_to_url(dep_v2_path)],
    )
    res = script.pip('install', pkga_path)
    assert "Successfully installed dep-2.0" in str(res), str(res)

    res = script.pip('install', pkga_path)
    return res


def test_install_pep508_with_url_in_install_requires_url_change_wheel(script):
    dep_v1_path = create_basic_wheel_for_package(
        script, name='dep', version='1.0',
    )

    res = function_that_does_a_lot(script, create_basic_wheel_for_package, dep_v1_path)

    # pip can determine the version from a wheel's filename, so the
    # dependency is not reinstalled if the URL doesn't change
    assert "Requirement already satisfied: dep==2.0" in str(res), str(res)


def test_install_pep508_with_url_in_install_requires_url_change_directory(
        script):
    dep_v1_path = create_test_package_with_setup(
        script, name='dep', version='1.0',
    )

    # Rename the package directory so it doesn't get overwritten when
    # creating the package for dep_v2
    dep_v1_path.move(dep_v1_path.folder / 'dep_v1')
    dep_v1_path = dep_v1_path.folder / 'dep_v1'

    res = function_that_does_a_lot(script, create_test_package_with_setup, dep_v1_path)

    # pip can't determine versions from a directory name, so it will always
    # reinstall the dependency
    assert "Successfully installed dep-2.0" in str(res), str(res)
```

If you have ideas to improve this I'm open to suggestions because I don't see a version of this that's better than the parametrization.

Reply:
Conceptually, it seems a lot simpler to me. Any way it's written, the reader will need to develop an understanding of what that large chunk of code is doing. How about starting by describing to me in words what the function is doing? (Not line by line but the goal in terms of setting up the test.)

Reply:
Also, something else to consider is that if you make your `create_dep` function in the second case one that returns a path dependent on the version (i.e. doing a move in both cases), then it looks like you can pass only `create_dep`, which would move more code into the common function. Aside from the function definition in the second case, the body of both test functions would be just two lines.